File: interviews/SAP/Garden.py
import sys 

class Garden:

    def canArrange(self, board, s, h): 

        m = len(board)
        n = len(board[0])

        row = [0 for _ in range(m)]
        col = [0 for _ in range(n)] 
        flowers = 0

        for i in range(m):
            for j in range(n): 
                if board[i][j] == "*":
                    row[i] += 1
                    col[j] += 1 
                    flowers += 1

        rowCap = flowers // (s + 1) 
        rowPart = 0 
        rowCount = s + 1
        rowArrange = False 

        for num in row: 

            rowPart += num  

            if rowPart < rowCap: 
                continue 
            elif rowPart == rowCap:
                rowPart = 0
                rowCount -= 1 

            else:
                break 

        if rowPart == 0 and rowCount == 0:
            rowArrange = True 

        else: 
            rowArrange = False 

        colCap = flowers // (h + 1) 
        colCount = h + 1
        colPart = 0
        colArrange = False 

        for num in col:

            colPart += num 

            if colPart < colCap:
                continue 
            elif colPart == colCap:
                colPart = 0
                colCount -= 1 

            else:
                break 

        if colPart == 0 and colCount == 0:
            colArrange = True 
        else:
            colArrange = False 

        return rowArrange and colArrange


garden = Garden() 
board = []
for line in sys.stdin: 

    line = line.rstrip()
    if line == "done":
        break
    board.append(line.split())

s = 0
h = 0 

print("Please input s")
for line in sys.stdin:
    
    line = line.rstrip()
    s = int(line)
    break 

print("Please input h")
for line in sys.stdin:
    
    line = line.rstrip()
    h = int(line)
    break 

print(garden.canArrange(board, 1, 1)) 

File: interviews/SAP/Garden.java
import java.io.*;
import java.util.*; 

public class Garden {

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in); 
        int n = sc.nextInt();
        System.out.println(n);

        String s = sc.next();
        sc.close();
        System.out.println(s);

        char[][] field = {{'*', '*', '*', '*'}, {'.', '.', '.', '*'}, {'*', '*', '.', '*'}};

        Garden garden = new Garden(); 
        System.out.println(garden.canArrange(field, 1, 1));

    }

    public boolean canArrange(char[][] field, int s, int h) {

        int[] row = new int[field.length]; 
        int[] col = new int[field[0].length]; 
        int flowers = 0; 
        for (int i = 0; i < field.length; i++) {
            for (int j = 0; j < field[0].length; j++) {

                if (field[i][j] == '*') {
                    row[i] += 1;
                    col[j] += 1;
                    flowers += 1;
                }
            }
        }
        
        int rowPart = 0;
        int rowCount = s + 1;
        int rowAverage = flowers / (s + 1);
        boolean rowArrange;

        for (int i = 0; i < row.length; i++) {
            
            rowPart += row[i]; 

            if (rowPart < rowAverage) {
                continue;
            }
            else if (rowPart == rowAverage) {
                rowCount -= 1;
                rowPart = 0;
            }
            else {
                rowArrange = false;
            }
        }

        if (rowPart == 0 && rowCount == 0) {
            rowArrange = true;
        }else {
            rowArrange = false;
        }

        int colPart = 0;
        int colCount = h + 1; 
        int colAverage = flowers / (h + 1);
        boolean colArrange; 

        for (int j = 0; j < col.length; j++) {

            colPart += col[j]; 

            if (colPart < colAverage) {
                continue;
            }
            else if (colPart == colAverage) {
                colCount -= 1;
                colPart = 0;
            }
            else {
                colArrange = false;
            }
        }

        if (colPart == 0 && colCount == 0) {
            colArrange = true;
        }
        else {
            colArrange = false;
        }

        return rowArrange && colArrange;

    }
}